File: packages-gen/js/libs/xt-net/net/conn-postgres.js
const Postgres = require("pg")

const xtd = require("@xtalk/lang/common-data.js")

const protocol = require("@xtalk/lang/common-protocol.js")

const conn_sql = require("@xtalk/net/conn-sql.js")

function PostgresClient(defaults,raw){
  globalThis["xt_lang_common_protocol$$IMPLEMENTATIONS"]["js.net.conn_postgres/PostgresClient"] = true;
  protocol.register_protocol_impl(conn_sql.ISqlClient["on"],"js.net.conn_postgres/PostgresClient",{
    "connect":client_connect,
    "disconnect":client_disconnect,
    "query":function (client,input){
        throw "Not Allowed";
      },
    "query_async":client_query_async
  });
  return {
    "::":"js.net.conn_postgres/PostgresClient",
    "::/protocols":[conn_sql.ISqlClient["on"]],
    "::/protocol-impls":{
        [conn_sql.ISqlClient["on"]]:{
          "connect":client_connect,
          "disconnect":client_disconnect,
          "query":function (client,input){
            throw "Not Allowed";
          },
          "query_async":client_query_async
        }
      },
    "defaults":defaults,
    "raw":raw
  };
}

function coerce_number_string(value){
  if(!("string" == (typeof value))){
    return value;
  }
  let trimmed = value.trim();
  if(trimmed == ""){
    return value;
  }
  if(trimmed.match("^[+-]?(?:\\d+(?:\\.\\d+)?|\\.\\d+)(?:[eE][+-]?\\d+)?$")){
    return Number(trimmed);
  }
  else{
    return value;
  }
}

function normalise_scalar_output(value){
  if((null == value) || ("boolean" == (typeof value)) || Array.isArray(value) || ((null != value) && ("object" == (typeof value)) && !Array.isArray(value))){
    return value;
  }
  else if("string" == (typeof value)){
    return coerce_number_string(value);
  }
  else{
    return value;
  }
}

function normalise_query_output(res){
  let {rows} = res;
  if((1 == rows.length) && (1 == xtd.obj_keys(xtd.first(rows)).length)){
    return normalise_scalar_output(xtd.obj_first_val(xtd.first(rows)));
  }
  else{
    return rows;
  }
}

function client_connect(client,opts){
  let {defaults} = client;
  let conn = new Postgres.Client(Object.assign(Object.assign({
    "host":"127.0.0.1",
    "port":5432,
    "user":"postgres",
    "password":"postgres",
    "database":"postgres"
  },defaults),opts));
  return conn.connect().then(function (){
    client["raw"] = conn;
    return client;
  });
}

function client_disconnect(client){
  let {raw} = client;
  raw.end();
  delete(client["raw"]);
  return client;
}

function client_query_async(client,input){
  let {raw} = client;
  return raw.query(input).then(normalise_query_output);
}

function create(defaults){
  return PostgresClient(defaults,null);
}

module.exports = {
  ["coerce_number_string"]:coerce_number_string,
  ["normalise_scalar_output"]:normalise_scalar_output,
  ["normalise_query_output"]:normalise_query_output,
  ["client_connect"]:client_connect,
  ["client_disconnect"]:client_disconnect,
  ["client_query_async"]:client_query_async,
  ["PostgresClient"]:PostgresClient,
  ["create"]:create
}

File: packages-gen/dart/libs/xt-net/lib/net/conn-postgres.dart
import 'package:postgres/postgres.dart' as pg;
import 'package:xtalk_lang/common-data.dart' as xtd;
import 'package:xtalk_net/conn-sql.dart' as conn_sql;
import 'dart:async';




DartPostgresClient(defaults, raw) {
  (__globals__["xt.lang.common_protocol/IMPLEMENTATIONS"] ??= <dynamic, dynamic>{})["dart.net.conn_postgres/DartPostgresClient"] = true;
  xt.lang.common_protocol.register_protocol_impl(conn_sql.ISqlClient["on"],"dart.net.conn_postgres/DartPostgresClient",<dynamic, dynamic>{
    "connect":client_connect,
    "disconnect":client_disconnect,
    "query":client_query,
    "query_async":client_query_async
  });
  return <dynamic, dynamic>{
    "::":"dart.net.conn_postgres/DartPostgresClient",
    "::/protocols":<dynamic>[conn_sql.ISqlClient["on"]],
    "::/protocol-impls":<dynamic, dynamic>{
        conn_sql.ISqlClient["on"]:<dynamic, dynamic>{
          "connect":client_connect,
          "disconnect":client_disconnect,
          "query":client_query,
          "query_async":client_query_async
        }
      },
    "defaults":defaults,
    "raw":raw
  };
}

default_env() {
  return <dynamic, dynamic>{
    "host":"127.0.0.1",
    "port":5432,
    "user":"postgres",
    "password":"postgres",
    "database":"test"
  };
}

normalise_scalar_output(value) {
  if("String" == (value.runtimeType).toString()){
    var parsed = num.tryParse( value );
    return parsed ?? value;
  }
  return value;
}

normalise_query_output(rows) {
  if(0 == rows.length){
    return <dynamic>[];
  }
  else if((1 == rows.length) && (1 == rows[0].length)){
    return normalise_scalar_output(rows[0][0]);
  }
  else{
    return xtd.arr_map(rows,(row) {
      return row.toColumnMap();
    });
  }
}

client_connect(client, opts) {
  var defaults = client["defaults"];
  var env = xtd.obj_clone(default_env());
  xtd.obj_assign(env,defaults);
  xtd.obj_assign(env,opts ?? <dynamic, dynamic>{});
  var endpoint = pg.Endpoint.new(
    host:env["host"],
    port:env["port"],
    database:env["database"],
    username:env["user"],
    password:env["password"]
  );
  var settings = pg.ConnectionSettings(connectTimeout: const Duration(seconds: 5), sslMode: pg.SslMode.disable);
  return ((Future.sync(() => pg.Connection.open(endpoint,settings:settings))) as Future<dynamic>).then((value) async { return await Function.apply((session) {
    client["raw"] = session;
    return client;
  },<dynamic>[value]); });
}

client_disconnect(client) {
  var raw = client["raw"];
  raw.close();
  return true;
}

client_query(client, query) {
  var raw = client["raw"];
  return ((Future.sync(() => raw.execute(query))) as Future<dynamic>).then((value) async { return await Function.apply((rows) {
    return normalise_query_output(rows);
  },<dynamic>[value]); });
}

client_query_async(client, query) {
  return client_query(client,query);
}

create(defaults) {
  return DartPostgresClient(defaults ?? <dynamic, dynamic>{},null);
}